packages/wrangler/src/assets.ts

```diff
@@ -538,18 +519,65 @@ export function validateAssetsArgsAndConfig(
 
 	// Smart placement turned on when using assets
 	if (
-		config?.placement?.mode === "smart" &&
-		config?.assets?.run_worker_first === true
+		config.placement?.mode === "smart" &&
+		config.assets?.run_worker_first === true
 	) {
 		logger.warn(
 			"Turning on Smart Placement in a Worker that is using assets and run_worker_first set to true means that your entire Worker could be moved to run closer to your data source, and all requests will go to that Worker before serving assets.\n" +
 				"This could result in poor performance as round trip times could increase when serving assets.\n\n" +
 				"Read more: https://developers.cloudflare.com/workers/static-assets/binding/#smart-placement"
 		);
 	}
+
+	if (
+		(args.assets || config.assets?.directory) &&
+		!(args.script || config.main)
+	) {
+		const unsupportedConfigKeys = getConfigKeysUnsupportedByAssetsOnly(config);
+
+		if (unsupportedConfigKeys.length > 0) {
+			const keys = unsupportedConfigKeys.map((key) => `⋅ "${key}"`).join("\n");
+
+			throw new UserError(
+				`Assets-only Workers do not support the following configuration keys:\n\n` +
+					`${keys}\n\n` +
+					`Please remove these fields from your configuration file, or configure the "main" field if you are trying to deploy a Worker with assets.`
+			);
+		}
+	}
 }
 
-const WORKER_JS_FILENAME = "_worker.js";
+function getConfigKeysUnsupportedByAssetsOnly(config: Config): Array<string> {
+	const supportedAssetsOnlyConfigKeys = new Set([
```

Comment:
@GregBrimble what do we wanna do abt args here? Some config fields, such as `vars` for example, can also be configured via flags (see `--var`). What's the expected behaviour we're looking for? IMHO it's kinda weird that configuring `[vars]` throws, while `--var` does not. At the same time, we need to consider that this function runs for 3 diff commands `dev`, `deploy`, and `versions upload`, so that's a pretty long list of flags we need to check against. The list of whitelisted config keys is very small, but that won't be the case for args, unless we can use some sort of subset?

Reply:
I think, it's kinda what Samuel was getting at earlier in a different comment: why isn't this function just receiving a fully-resolved config object? By this point, surely, there's nothing that assets validation cares about which would be specific to args or config: it only cares about which features are enabled/disabled. Whether an arg was used or config was used to enable something that isn't supported, it doesn't matter.

Is there any sane way we can combine them together before calling this?

Reply:
https://github.com/cloudflare/workers-sdk/pull/8998/files#r2068920006